nixos/modules/services/torrent/qbittorrent.nix

```diff
@@ -0,0 +1,150 @@
+  gendeepINI = toINI {
+    mkKeyValue = let
+      sep = "=";
+    in
+      k: v:
+        if isAttrs v
+        then
+          concatStringsSep "\n"
+          (mapAttrsToList (k2: v2:
+            "${escape [sep] "${k}\\${k2}"}${sep}${mkValueStringDefault {} v2}"
+            )
+          v)
+        else mkKeyValueDefault {} sep k v;
+  };
+in
+{
+  options = {
+    services.qbittorrent = {
+      enable = mkEnableOption (mdDoc "qbittorrent BitTorrent client");
+
+      user = mkOption {
+        type = str;
+        default = "qbittorrent";
+        description = mdDoc "User account under which qBitTorrent runs.";
+      };
+
+      group = mkOption {
+        type = str;
+        default = "qbittorrent";
+        description = mdDoc "Group under which qBitTorrent runs.";
+      };
+
+      package = mkPackageOption pkgs "qbittorrent-nox" { };
+
+      extraPackages = mkOption {
+        type = listOf package;
+        description = mdDoc ''
+          Extra packages to be installed system wide. See
+          https://github.com/qbittorrent/qBittorrent/wiki/List-of-known-alternate-WebUIs
+        '';
+      };
+
+      profileDir = mkOption {
+        type = str;
+        default = "/var/lib/qBittorrent/";
+        description = mdDoc "the path passed to qbittorrent via --profile";
+      };
+
+      serverConfig = mkOption {
+        type = unspecified;
+      };
+
+      openFirewall = mkOption {
+        type = bool;
+        default = false;
+        example = true;
+        description = mdDoc ''
+          Opens both ports under `serverConfig.BitTorrent.Session.Port` and `serverConfig.Preferences.WebUI.Port`,
+          if the former is not specified then no port is opened as the default is random. when the latter is not specified the default of `8080` is opened
+        '';
+      };
+    };
+  };
+  config = let
+    conf = pkgs.writeText "qBittorrent.conf" (gendeepINI cfg.serverConfig);
+  in lib.mkIf cfg.enable {
+    systemd = {
+      tmpfiles.settings = {
+        profileDir = {
+          "${cfg.profileDir}"."d" = {
+            mode = "0755";
+            user = cfg.user;
+            group = cfg.group;
+          };
+        };
+        configFile = {
+          "${cfg.profileDir}/qBittorrent/"."d" = {
+            mode = "0755";
+            user = cfg.user;
+            group = cfg.group;
+          };
+          "${cfg.profileDir}/qBittorrent/config/"."d" = {
+            mode = "0755";
+            user = cfg.user;
+            group = cfg.group;
+          };
+          "${cfg.profileDir}/qBittorrent/config/qBittorrent.conf"."L+" = {
+            mode = "0644";
+            user = cfg.user;
+            group = cfg.group;
+            argument = "${conf}";
+          };
+        };
+      };
+      # following https://github.com/qbittorrent/qBittorrent/blob/master/dist/unix/systemd/qbittorrent-nox%40.service.in
+      services.qbittorrent = {
+        description = "qbittorrent BitTorrent client";
```

Comment:
The qbittorrent-nox package already includes a systemd service, is it necessary to replicate it?

Reply:
to enable the ability to easily define which package is used for this module I believe so yeah. though I was unaware it had a service so I'll check it out regardless.

another thing is the definition of its working directory and the `--profile` option on the run line. it seems cleaner but if there's a nicer way to use the existing service while keeping these options please let me know